File: my-folder/1558-course-schedule-iv/solution.py
class Solution:
    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:

        graph = defaultdict(list)
        queue = deque()

        dp = [[0]*numCourses for _ in range(numCourses)]

        for p, c in prerequisites:
            dp[c][p] = 1
            graph[c].append(p)
    
        for i in range(numCourses):
            queue.append(i)
            visited = set()
            visited.add(i)

            while queue:
                course = queue.popleft()
                for prereq in graph[course]:
                    if prereq not in visited:
                        queue.append(prereq)
                        dp[i][prereq] = 1
                        visited.add(prereq)
        print(dp)
        ans = []
        for p, c in queries:
            ans.append(dp[c][p]==1)

        return ans

            

File: my-folder/1558-course-schedule-iv/solution.java
class Solution {
    public List<Boolean> checkIfPrerequisite(int numCourses, int[][] prerequisites, int[][] queries) {
        
        Map<Integer, List<Integer>> graph = new HashMap<>();
        Queue<Integer> queue = new LinkedList<>();
        int[][] dp = new int[numCourses][numCourses];

        for(int i=0; i < prerequisites.length; i++){
            int p = prerequisites[i][0];
            int c = prerequisites[i][1];
            graph.putIfAbsent(c, new ArrayList<>());
            graph.get(c).add(p);
            dp[c][p] = 1;
        }

        for (int i=0; i<numCourses; i++){
            queue.offer(i);
            Set<Integer> visited = new HashSet<>();
            visited.add(i);

            while (!queue.isEmpty()){
                int course = queue.poll();
                if (graph.containsKey(course)){
                    List<Integer> prereqs = graph.get(course);
                    for(int j=0; j<prereqs.size(); j++){
                        int prereq = prereqs.get(j);
                        if (!visited.contains(prereq)){
                            queue.offer(prereq);
                            dp[i][prereq] = 1;
                            visited.add(prereq);
                        }
                    }
                }
            }
        }

        List<Boolean> ans = new ArrayList<>(queries.length);

        for (int i=0; i<queries.length; i++){
            int p = queries[i][0];
            int c = queries[i][1];
            if (dp[c][p] == 1){
                ans.add(true);
            }else{
                ans.add(false);
            }
            
        }

        return ans;
    }
}
